xchange-therock/src/main/java/org/knowm/xchange/therock/dto/marketdata/TheRockTicker.java

```diff
@@ -3,21 +3,71 @@
 import com.fasterxml.jackson.databind.PropertyNamingStrategy;
 import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
 import com.fasterxml.jackson.databind.annotation.JsonNaming;
+import java.io.Serializable;
 import java.math.BigDecimal;
 import java.util.Date;
 import org.knowm.xchange.currency.CurrencyPair;
-import org.knowm.xchange.utils.jackson.CurrencyPairDeserializer;
+import org.knowm.xchange.therock.utils.jackson.TheRockTickerDeserializer;
 
 /** @author Matija Mazi */
 @JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy.class)
-public class TheRockTicker {
+@JsonDeserialize(using = TheRockTickerDeserializer.class)
```

Comment:
first of all is this needed here for the Instrument refactor to work?

secondly you should not write deserializer for DTOs that can be easily deserialized by Jackson it self if you just use Jacksons annotations properly

Reply:
Hi Adam,

The issue I was struggling with, if we leave it as is we have:

`@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy.class)
public class TheRockTicker {

  @JsonDeserialize(using = CurrencyPairDeserializer.class)
  private CurrencyPair fundId;`

But when we go to deserialize it, jackson cannot work out the type.

`com.fasterxml.jackson.databind.exc.InvalidTypeIdException: 
Missing type id when trying to resolve subtype of [simple type, class org.knowm.xchange.currency.CurrencyPair]: missing type id property 'type' (for POJO property 'fund_id')`

Thought the `JsonDeserialize` annotation would have resolved the type, but it did not, suggestions welcomed on which annotation might work.

Reply:
yes and that is because these annotations:

```
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
@JsonSubTypes({@Type(value = CurrencyPair.class, name = "currencyPair")})
```

on Instrument currently make no sense if we serialize / deserialize CurrencyPair as a string 

a string will never have a property `type` - we will probably have to have a custom deserializer for Insturment to determine its concrete class

I see you have already provided one `InstrumentDeserializer` - which is better then having a custom deserializer for every DTO that uses the Insturment class - this would be a show stopper for this refactor

for now remove those annotation

we also must think how will we serialize / deserialize derivatives

in your `InstrumentDeserializer` you detect it the string is currency pair based by the number of `/` characters - that is ok

but how will we distinguish futures from options? could you give some examples how their ticker symbols look on exchanges?

Reply:
Thanks so much for suggestions, yeah a customer deserializer seems the way to go.
It is not ideal, but deserialising from strings we can count the `/`
CurrencyPair (Base`/`Counter)
BTC`/`USD
Futures/Swaps (Base`/`Counter`/`Prompt)
BTC`/`USD`/`200925
Options (Base`/`Counter`/`Prompt`/`StrikePrice`/`Put?Call)
BTC`/`USD`/`200925`/`8956.67`/`P